File: backend/app/routers/webhooks.py
"""Webhook endpoints for provider callbacks."""

from __future__ import annotations

from contextlib import asynccontextmanager
from typing import Any, Mapping, cast

from fastapi import APIRouter, Depends, HTTPException, Request, status
from pydantic import BaseModel

from ..billing import stripe_service
from ..db import get_db

router = APIRouter(prefix="/webhooks", tags=["webhooks"])


class StripeWebhookResponse(BaseModel):
    status: str
    duplicate: bool = False


def _record_to_dict(record: Any | None) -> dict[str, Any] | None:
    if record is None:
        return None
    if isinstance(record, Mapping):
        return dict(record)
    return cast(dict[str, Any], dict(record))


@asynccontextmanager
async def transaction_scope(db: Any):
    transaction = getattr(db, "transaction", None)
    if transaction is None:
        yield
        return

    async with transaction():
        yield


async def fetch_logged_event(
    db: Any,
    stripe_event_id: str,
    *,
    for_update: bool = False,
) -> dict[str, Any] | None:
    lock_clause = " FOR UPDATE" if for_update else ""
    row = await db.fetchrow(
        f"""
        SELECT stripe_event_id, processed_at
        FROM stripe_events
        WHERE stripe_event_id = $1
        {lock_clause}
        """,
        stripe_event_id,
    )
    return _record_to_dict(row)


async def insert_logged_event(
    db: Any,
    *,
    stripe_event_id: str,
    event_type: str,
    payload: str,
) -> dict[str, Any] | None:
    row = await db.fetchrow(
        """
        INSERT INTO stripe_events (stripe_event_id, event_type, payload)
        VALUES ($1, $2, $3)
        ON CONFLICT (stripe_event_id) DO NOTHING
        RETURNING stripe_event_id, processed_at
        """,
        stripe_event_id,
        event_type,
        payload,
    )
    return _record_to_dict(row)


async def mark_event_processed(db: Any, stripe_event_id: str) -> None:
    await db.execute(
        """
        UPDATE stripe_events
        SET processed_at = NOW()
        WHERE stripe_event_id = $1
        """,
        stripe_event_id,
    )


async def process_stripe_event(db: Any, event: Mapping[str, Any]) -> None:
    event_type = str(event.get("type") or "")
    event_object = cast(
        dict[str, Any],
        ((event.get("data") or {}).get("object") or {}),
    )

    if event_type == "checkout.session.completed":
        metadata = cast(dict[str, Any], event_object.get("metadata") or {})
        user_id = cast(str | None, metadata.get("user_id") or event_object.get("client_reference_id"))
        if not user_id:
            return

        await stripe_service.activate_checkout_subscription(
            db,
            user_id=user_id,
            stripe_customer_id=cast(str | None, event_object.get("customer")),
            subscription_id=cast(str | None, event_object.get("subscription")),
        )
        return

    if event_type in {"customer.subscription.deleted", "customer.subscription.updated"}:
        stripe_customer_id = cast(str | None, event_object.get("customer"))
        if not stripe_customer_id:
            return

        await stripe_service.sync_subscription_status(
            db,
            stripe_customer_id,
            event_object,
        )


@router.post("/stripe", response_model=StripeWebhookResponse)
async def handle_stripe_webhook(
    request: Request,
    db: Any = Depends(get_db),
) -> StripeWebhookResponse:
    payload = await request.body()

    try:
        event = stripe_service.construct_webhook_event(
            payload,
            request.headers.get("Stripe-Signature"),
        )
    except stripe_service.StripeWebhookVerificationError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(exc),
        ) from exc

    stripe_event_id = cast(str | None, event.get("id"))
    event_type = cast(str | None, event.get("type"))
    if not stripe_event_id or not event_type:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Stripe event is missing required identifiers.",
        )

    existing = await fetch_logged_event(db, stripe_event_id)
    if existing and existing.get("processed_at") is not None:
        return StripeWebhookResponse(status="duplicate", duplicate=True)

    async with transaction_scope(db):
        inserted = await insert_logged_event(
            db,
            stripe_event_id=stripe_event_id,
            event_type=event_type,
            payload=payload.decode("utf-8"),
        )
        if inserted is None:
            locked_event = await fetch_logged_event(
                db,
                stripe_event_id,
                for_update=True,
            )
            if locked_event and locked_event.get("processed_at") is not None:
                return StripeWebhookResponse(status="duplicate", duplicate=True)

        try:
            await process_stripe_event(db, event)
        except stripe_service.StripeServiceError as exc:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=str(exc),
            ) from exc

        await mark_event_processed(db, stripe_event_id)

    return StripeWebhookResponse(status="ok", duplicate=False)

File: api/src/routes/webhooks.ts
import { Hono } from "hono";

import type { DatabaseClient } from "../db/client";
import {
  fulfillAutoRechargePayment,
  fulfillTopupCheckout,
  fulfillSubscriptionInvoice,
  recordFailedInvoice,
  reverseBillingOrderByPaymentIntent
} from "../services/billing";
import {
  activateCheckoutSubscription,
  constructWebhookEvent,
  StripeServiceError,
  StripeWebhookVerificationError,
  syncSubscriptionStatus
} from "../services/stripe";
import { sendBillingNotification } from "../services/transactional-email";
import { apiError } from "../utils/http";

async function fetchLoggedEvent(db: DatabaseClient, stripeEventId: string): Promise<Record<string, unknown> | null> {
  return db.fetchrow(
    `
      SELECT stripe_event_id, processed_at
      FROM stripe_events
      WHERE stripe_event_id = $1
    `,
    stripeEventId
  );
}

async function insertLoggedEvent(
  db: DatabaseClient,
  stripeEventId: string,
  eventType: string,
  payload: string
): Promise<Record<string, unknown> | null> {
  return db.fetchrow(
    `
      INSERT INTO stripe_events (stripe_event_id, event_type, payload)
      VALUES ($1, $2, $3)
      ON CONFLICT (stripe_event_id) DO NOTHING
      RETURNING stripe_event_id, processed_at
    `,
    stripeEventId,
    eventType,
    payload
  );
}

async function markEventProcessed(db: DatabaseClient, stripeEventId: string): Promise<void> {
  await db.execute(
    `
      UPDATE stripe_events
      SET processed_at = NOW()
      WHERE stripe_event_id = $1
    `,
    stripeEventId
  );
}

function asRecord(value: unknown): Record<string, unknown> {
  return typeof value === "object" && value != null ? value as Record<string, unknown> : {};
}

function asString(value: unknown): string | null {
  return typeof value === "string" && value.trim() ? value.trim() : null;
}

function asInteger(value: unknown): number {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? Math.max(Math.trunc(parsed), 0) : 0;
}

function stripeCreatedAt(value: unknown): Date | null {
  const parsed = Number(value);
  if (!Number.isFinite(parsed)) {
    return null;
  }
  return new Date(parsed * 1000);
}

function sumDiscounts(value: unknown): number {
  if (!Array.isArray(value)) {
    return 0;
  }
  return value.reduce((sum, item) => sum + asInteger(asRecord(item).amount), 0);
}

function extractInvoicePeriod(invoice: Record<string, unknown>): { periodStart: string; periodEnd: string } {
  const lines = Array.isArray(asRecord(invoice.lines).data) ? asRecord(invoice.lines).data as unknown[] : [];
  const firstLine = lines.find((line) => typeof line === "object" && line != null) as Record<string, unknown> | undefined;
  const period = firstLine ? asRecord(firstLine.period) : {};
  const startSeconds = Number(period.start);
  const endSeconds = Number(period.end);
  const createdAt = stripeCreatedAt(invoice.created) ?? new Date();

  if (Number.isFinite(startSeconds) && Number.isFinite(endSeconds)) {
    const periodStart = new Date(startSeconds * 1000).toISOString().slice(0, 10);
    const periodEnd = new Date(Math.max(endSeconds * 1000 - 1000, startSeconds * 1000)).toISOString().slice(0, 10);
    return { periodStart, periodEnd };
  }

  const year = createdAt.getUTCFullYear();
  const month = createdAt.getUTCMonth();
  const periodStart = new Date(Date.UTC(year, month, 1)).toISOString().slice(0, 10);
  const periodEnd = new Date(Date.UTC(year, month + 1, 0)).toISOString().slice(0, 10);
  return { periodStart, periodEnd };
}

function isCheckoutComplete(checkoutSession: Record<string, unknown>): boolean {
  return asString(checkoutSession.status) === "complete";
}

function isPaymentCheckoutPaid(checkoutSession: Record<string, unknown>): boolean {
  return asString(checkoutSession.payment_status) === "paid";
}

function isSubscriptionCheckoutReady(checkoutSession: Record<string, unknown>): boolean {
  const paymentStatus = asString(checkoutSession.payment_status);
  return paymentStatus === "paid" || paymentStatus === "no_payment_required";
}

async function processStripeEvent(db: DatabaseClient, config: any, event: Record<string, unknown>): Promise<void> {
  const eventType = String(event.type ?? "");
  const eventObject = asRecord(asRecord(event.data).object);

  if (eventType === "checkout.session.completed" || eventType === "checkout.session.async_payment_succeeded") {
    const metadata = asRecord(eventObject.metadata);
    const mode = asString(eventObject.mode) ?? "";
    if (mode === "setup") {
      if (!isCheckoutComplete(eventObject)) {
        return;
      }
      const stripeCustomerId = asString(eventObject.customer);
      if (!stripeCustomerId) {
        return;
      }
      await db.execute(
        `
          UPDATE user_profiles
          SET
              has_payment_method_on_file = TRUE,
              updated_at = NOW()
          WHERE stripe_customer_id = $1
        `,
        stripeCustomerId
      );
      return;
    }

    const userId = asString(metadata.user_id) ?? asString(eventObject.client_reference_id);
    if (!userId) {
      return;
    }

    if (mode === "subscription") {
      if (!isCheckoutComplete(eventObject) || !isSubscriptionCheckoutReady(eventObject)) {
        return;
      }
      await activateCheckoutSubscription(
        db,
        userId,
        asString(eventObject.customer),
        asString(eventObject.subscription)
      );
      await db.execute(
        `
          UPDATE user_profiles
          SET
              has_payment_method_on_file = TRUE,
              updated_at = NOW()
          WHERE id = $1
        `,
        userId
      );
      return;
    }

    if (mode === "payment") {
      if (!isPaymentCheckoutPaid(eventObject)) {
        return;
      }
      const quantity = Number(metadata.quantity) || 1000;
      const notification = await fulfillTopupCheckout(db, {
        userId,
        credits: quantity,
        stripeCheckoutSessionId: String(eventObject.id ?? ""),
        stripeCustomerId: asString(eventObject.customer),
        stripePaymentIntentId: asString(eventObject.payment_intent),
        currency: asString(eventObject.currency),
        grossAmountCents: asInteger(eventObject.amount_subtotal ?? eventObject.amount_total),
        discountAmountCents: asInteger(asRecord(eventObject.total_details).amount_discount),
        netAmountCents: asInteger(eventObject.amount_total),
        occurredAt: stripeCreatedAt(eventObject.created)
      });
      if (notification) {
        void sendBillingNotification(config, notification).catch((error) => {
          console.error("[billing] Failed to send top-up email:", error);
        });
      }
    }
    return;
  }

  if (eventType === "invoice.paid") {
    const stripeCustomerId = asString(eventObject.customer);
    if (!stripeCustomerId) {
      return;
    }
    await db.execute(
      `
        UPDATE user_profiles
        SET
            has_payment_method_on_file = TRUE,
            updated_at = NOW()
        WHERE stripe_customer_id = $1
      `,
      stripeCustomerId
    );
    await syncSubscriptionStatus(db, stripeCustomerId, {
      id: asString(eventObject.subscription),
      status: "active"
    });
    const { periodStart, periodEnd } = extractInvoicePeriod(eventObject);
    const notification = await fulfillSubscriptionInvoice(db, {
      stripeInvoiceId: String(eventObject.id ?? ""),
      stripeCustomerId,
      stripeSubscriptionId: asString(eventObject.subscription),
      stripePaymentIntentId: asString(eventObject.payment_intent),
      currency: asString(eventObject.currency),
      grossAmountCents: asInteger(eventObject.subtotal ?? eventObject.amount_due ?? eventObject.amount_paid),
      discountAmountCents: sumDiscounts(eventObject.total_discount_amounts),
      netAmountCents: asInteger(eventObject.amount_paid ?? eventObject.amount_due),
      periodStart,
      periodEnd,
      occurredAt: stripeCreatedAt(eventObject.created),
      metadata: {
        billing_reason: asString(eventObject.billing_reason)
      }
    });
    if (notification) {
      void sendBillingNotification(config, notification).catch((error) => {
        console.error("[billing] Failed to send subscription email:", error);
      });
    }
    return;
  }

  if (eventType === "invoice.payment_failed") {
    const stripeCustomerId = asString(eventObject.customer);
    if (!stripeCustomerId) {
      return;
    }
    await recordFailedInvoice(db, {
      stripeInvoiceId: String(eventObject.id ?? ""),
      stripeCustomerId,
      stripeSubscriptionId: asString(eventObject.subscription),
      stripePaymentIntentId: asString(eventObject.payment_intent),
      currency: asString(eventObject.currency),
      grossAmountCents: asInteger(eventObject.subtotal ?? eventObject.amount_due),
      discountAmountCents: sumDiscounts(eventObject.total_discount_amounts),
      netAmountCents: asInteger(eventObject.amount_due),
      metadata: {
        billing_reason: asString(eventObject.billing_reason)
      }
    });
    return;
  }

  if (eventType === "customer.subscription.deleted" || eventType === "customer.subscription.updated") {
    const stripeCustomerId = asString(eventObject.customer);
    if (!stripeCustomerId) {
      return;
    }
    await syncSubscriptionStatus(db, stripeCustomerId, eventObject);
    return;
  }

  if (eventType === "payment_intent.succeeded") {
    const metadata = asRecord(eventObject.metadata);
    if (asString(metadata.type) === "auto_recharge") {
      const userId = asString(metadata.user_id);
      const quantity = Number(metadata.quantity) || 1000;
      if (userId) {
        const amount = asInteger(eventObject.amount_received ?? eventObject.amount);
        const notification = await fulfillAutoRechargePayment(db, {
          userId,
          stripePaymentIntentId: String(eventObject.id ?? ""),
          stripeCustomerId: asString(eventObject.customer),
          quantity,
          currency: asString(eventObject.currency),
          grossAmountCents: amount,
          discountAmountCents: 0,
          netAmountCents: amount,
          occurredAt: stripeCreatedAt(eventObject.created)
        });
        if (notification) {
          void sendBillingNotification(config, notification).catch((error) => {
            console.error("[billing] Failed to send auto-recharge email:", error);
          });
        }
      }
    }
    return;
  }

  if (eventType === "charge.refunded") {
    const paymentIntentId = asString(eventObject.payment_intent);
    if (paymentIntentId) {
      await reverseBillingOrderByPaymentIntent(db, paymentIntentId, "refunded");
    }
    return;
  }

  if (eventType === "charge.dispute.created" || eventType === "charge.dispute.funds_withdrawn") {
    const paymentIntentId = asString(asRecord(eventObject.payment_intent).id) ?? asString(eventObject.payment_intent);
    if (paymentIntentId) {
      await reverseBillingOrderByPaymentIntent(db, paymentIntentId, "disputed");
    }
  }
}

export function createWebhookRouter(): any {
  const router = new Hono();

  router.post("/stripe", async (c: any) => {
    const db = c.get("db") as DatabaseClient;
    const config = c.get("config");
    const payload = await c.req.raw.text();
    let event: Record<string, unknown>;
    try {
      event = constructWebhookEvent(config, payload, c.req.header("stripe-signature") ?? null);
    } catch (error) {
      if (error instanceof StripeWebhookVerificationError) {
        apiError(400, error.message);
      }
      throw error;
    }

    const stripeEventId = String(event.id ?? "").trim();
    const eventType = String(event.type ?? "").trim();
    if (!stripeEventId || !eventType) {
      apiError(400, "Stripe event is missing required identifiers.");
    }

    const existing = await fetchLoggedEvent(db, stripeEventId);
    if (existing?.processed_at != null) {
      return c.json({ status: "duplicate", duplicate: true });
    }

    return db.transaction(async (tx) => {
      const inserted = await insertLoggedEvent(tx, stripeEventId, eventType, payload);
      if (inserted == null) {
        const lockedEvent = await tx.fetchrow(
          `SELECT stripe_event_id, processed_at FROM stripe_events WHERE stripe_event_id = $1 FOR UPDATE`,
          stripeEventId
        );
        if (lockedEvent?.processed_at != null) {
          return c.json({ status: "duplicate", duplicate: true });
        }
      }

      try {
        await processStripeEvent(tx, config, event);
      } catch (error) {
        if (error instanceof StripeServiceError) {
          apiError(409, error.message);
        }
        throw error;
      }

      await markEventProcessed(tx, stripeEventId);
      return c.json({ status: "ok", duplicate: false });
    });
  });

  return router;
}
